src/mesh/curved_mesh.jl

```diff
@@ -0,0 +1,170 @@
+"""
+    CurvedMesh{NDIMS, RealT<:Real} <: AbstractMesh{NDIMS}
+
+A structured curved mesh.
+
+Different numbers of cells per dimension are possible and arbitrary functions 
+can be used as domain faces.
+
+Note: This is an experimental feature and may be changed in future releases without notice.
+"""
+mutable struct CurvedMesh{NDIMS, RealT<:Real} <: AbstractMesh{NDIMS}
+  cells_per_dimension::NTuple{NDIMS, Int}
+  faces::Tuple
```

Comment:
We were thinking about this, but this would add another otherwise unused type parameter for nearly no benefit. The faces are only called once in the `init_elements!` function and never in any performance relevant code. Should we still add the parameter?

Reply:
Hm, you're right. Only add it if it is performance critical. @ranocha would this non fully specified type have negative impacts in locations where we only access other fields of the mesh? Eg, because now the mesh is not storable on the stack or because the compiler may not know the memory layout beforehand?

Reply:
I wouldn't expect to see huge performance differences. However, I would propose to use 
```julia
faces::Any # some comment why this is used
```
This makes it explicit that `faces` should not be fully specified.

Reply:
OK, sounds good to me. We'll see if `faces` will remain here once the mesh supports arbitrary curvature in the interior anyways, so I think a reasonable solution is sufficient here.